File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/outtake/Lift.java
package org.firstinspires.ftc.teamcode.outtake;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.util.Range;
import org.firstinspires.ftc.teamcode.utils.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.ManualPositionMechanism;
import org.jetbrains.annotations.NotNull;

import static org.firstinspires.ftc.teamcode.Positions.Outtake.Lift.*;

/**
 * Class representing the vertical movement of the outtake.
 * It implements the ManualPositionMechanism interface to control the lift motors.
 */
public class Lift implements ManualPositionMechanism {
    private boolean isCancelled = false;
    private final DcMotorEx[] motors;
    /**
     * Constructor for the Lift class.
     * Initializes the lift motors using the provided hardware map.
     * @param hardwareMap the hardware map to use for initialization.
     */
    public Lift(@NotNull HardwareMap hardwareMap) {
        motors = new DcMotorEx[]{
            hardwareMap.get(DcMotorEx.class, "LiftLeft"),
            hardwareMap.get(DcMotorEx.class, "LiftRight")
        };
        for (DcMotorEx motor : motors) {
            motor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
            motor.setMode(DcMotorEx.RunMode.STOP_AND_RESET_ENCODER);
            motor.setTargetPosition(0);
            motor.setMode(DcMotorEx.RunMode.RUN_USING_ENCODER);
        }
        motors[1].setDirection(DcMotorEx.Direction.REVERSE);
    }
    private double power = 1.0;
    private int targetPosition = 0;
    /**
     * Sets the target position for the lift motors.
     * @param position the target position to set.
     */
    @Override
    public void setTargetPosition(double position) {
        targetPosition = (int) Range.clip(position, 0, 2185);
    }
    /**
     * Gets the target position for the lift motors.
     * @return the target position.
     */
    @Override
    public double getTargetPosition() {
        return targetPosition;
    }
    private double multiplier = 15.0;
    /**
     * Adjusts the multiplier for the lift motors.
     * @param multiplier the multiplier to adjust by.
     */
    @Override
    public void adjustMultiplier(double multiplier) {
        this.multiplier = multiplier;
    }
    /**
     * Gets the multiplier for the lift motors.
     * @return the multiplier.
     */
    @Override
    public double getMultiplier() {
        return multiplier;
    }
    /**
     * Cancels the lift motors.
     */
    @Override
    public void cancel() {
        power = 0.0;
        isCancelled = true;
    }
    /**
     * Runs the lift motors and updates the telemetry packet.
     * @param packet the telemetry packet to update.
     * @return true if the motors are running, false otherwise.
     */
    @Override
    public boolean run(@NotNull TelemetryPacket packet) {
        if (isCancelled) {
            for (DcMotorEx motor : motors) motor.setPower(power);
            return false;
        }
        for(int ind = 0; ind < motors.length; ind++) {
            motors[ind].setPower(power);
            motors[ind].setTargetPosition(targetPosition);

            packet.put("Lift " + ind + " position", motors[ind].getCurrentPosition());
        }
        return true;
    }
}

/**
 * TeleOp class for testing the Lift mechanism manually.
 */
@TeleOp(name = "Lift Test", group = "C Outtake")
class LiftTest extends ManualMechanismTeleOp {
    public LiftTest() {
        super(Lift::new);
    }
}

/**
 * TeleOp class for testing the Lift mechanism positions.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Lift Positions Test", group = "D Outtake")
@Config
class LiftPositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Lift lift = new Lift(hardwareMap);
        waitForStart();
        while (opModeIsActive()) {
            if (gamepad1.dpad_down) current = down;
            else if (gamepad1.dpad_right) current = half;
            else if (gamepad1.dpad_up) current = up;
            else current += lift.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            lift.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if(!lift.run(packet)) requestOpModeStop();
            telemetry.addData("Position", current);
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/outtake/Lift.kt
package org.firstinspires.ftc.outtake

import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.utils.DelayedActions
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.ManualPositionMechanism
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.PressAction

@Config
class Lift(
    hardwareMap: HardwareMap,
    val motors: Array<DcMotorEx> = arrayOf(
        hardwareMap.get(DcMotorEx::class.java, "Lift Left"),
        hardwareMap.get(DcMotorEx::class.java, "Lift Right")
    )
) : ManualPositionMechanism {
    companion object {
        var lastKnown = 0
        /**
         * The up position of the lift.
         */
        @JvmField
        @Volatile
        var up = 3200.0
        /**
         * The position of the lift the level 2 hang is finished.
         */
        @JvmField
        @Volatile
        var hang = 1700.0
        /**
         * The position of the lift when placing specimens in the high chamber.
         */
        @JvmField
        @Volatile
        var half = 650.0
        /**
         * The position of the lift when it transfer.
         */
        @JvmField
        @Volatile
        var transfer = 100.0
        /**
         * The down position of the lift.
         */
        @JvmField
        @Volatile
        var down = 0.0
    }
    /**
     * Flag for whether the lift is cancelled.
     */
    private var isCancelled = false
    /**
     * Initializes the mechanism`s components.
     */
    init {
        motors.forEach {
            it.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
            it.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
            it.mode = DcMotor.RunMode.RUN_TO_POSITION
            it.power = 1.0
        }
        motors[0].direction = DcMotorSimple.Direction.REVERSE
    }
    /**
     * The current target position of the lift.
     */
    override var targetPosition: Double = 0.0
        set(value) {
            field = value.coerceIn(0.0..3200.0)
        }
    /**
     * The multiplier that the mechanism gets manually adjusted by.
     */
    override var adjustMultiplier: Double = 15.0
    /**
     * The current position of the lift.
     */
    val measuredPosition: Double
        get() = motors.map { it.currentPosition }.average()
    /**
     * The cancel function of the mechanism.
     * Also sets the power of the motors to 0 and stores the last known position of the lift for future use.
     */
    override fun cancel() {
        motors.forEach { it.power = 0.0 }
        lastKnown = measuredPosition.toInt()
        isCancelled = true
    }
    /**
     * The update function of the mechanism.
     */
    override fun run(p: TelemetryPacket): Boolean {
        if (isCancelled) {
            isCancelled = false
            return false
        }
        for ((ind, motor) in motors.withIndex()) {
            motor.targetPosition = targetPosition.toInt()
            p.put("Lift $ind position", motor.currentPosition)
        }
        return true
    }
}
/**
 * Autonomous class for resetting the lift.
 */
@Autonomous(name = "Reset Lift", group = "C")
@Config
private class ResetLift : LinearOpMode() {
    companion object {
        /**
         * The power at which the lift is reset.
         */
        private var resetPower = -0.2
    }
    override fun runOpMode() {
        val lift = Lift(hardwareMap)
        val delays = DelayedActions()
        lift.motors.forEach {
            it.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
            it.power = resetPower
        }
        delays.addDelay(3) {
            lift.motors.forEach {
                it.power = 0.0
                it.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
                it.mode = DcMotor.RunMode.RUN_TO_POSITION
            }
            lift.cancel()
            requestOpModeStop()
        }
        while (opModeIsActive()) {
            delays.run()
        }
    }
}
/**
 * TeleOp class for testing the Lift mechanism manually.
 */
@TeleOp(name = "Lift Test", group = "C")
@Disabled
private class LiftTest : ManualMechanismTeleOp(::Lift)
/**
 * TeleOp class for testing the positions of the Lift mechanism.
 */
@TeleOp(name = "Lift Position Test", group = "D")
@Disabled
private class LiftPositions : Movement() {
    /**
     * The Lift mechanism to be tested.
     */
    override lateinit var system: Lift

    override fun systemInit() {
        system = Lift(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_up) { system.targetPosition = Lift.up })
        actions.add(PressAction(gamepad1::dpad_left) { system.targetPosition = Lift.hang })
        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = Lift.half })
        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = Lift.down })
        actions.add(PressAction(gamepad1::cross) { system.targetPosition = Lift.transfer })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}
